### PR TITLE
WS-20/21 Adding warehouse controller

### PR DESCRIPTION
**Pull request by anto-belo**
PR contains implementation of both WS-20 (Create Warehouse management page) and WS-21 (Create Warehouse form) tickets

### CODE REVIEW

File: src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+public class WarehouseController {
+    private final WarehouseService warehouseService;
+
+    @PreAuthorize("hasAuthority('warehouses:read')")
+    @GetMapping("/warehouses")
+    ResponseEntity<List<WarehouseDto>> getByPage(
+            @RequestParam(required = false, defaultValue = "0") int page,
+            @RequestParam(required = false, defaultValue = "10") int size,
+            @NotNull(message = "Customer must be specified")
+            @RequestBody(required = false) CustomerDto customerDto) {
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java:36]: if it must not be null, why is it `required = false`?

File: src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        WarehouseDto warehouse = warehouseService.findById(id);
+        if (warehouse == null) {
+            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+        }
+        return new ResponseEntity<>(warehouse, HttpStatus.OK);
+    }
+
+    @PreAuthorize("hasAuthority('warehouses:write')")
+    @PostMapping("/warehouse")
+    ResponseEntity<WarehouseDto> create(@Valid @RequestBody WarehouseDto warehouseDto) {
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java:58]: Why responseEntity, just use `@ResponseStatus`

File: src/main/java/com/itechart/students_lab/waybill_suppliers/entity/dto/EmployeeDto.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package com.itechart.students_lab.waybill_suppliers.entity.dto;
+
+public class EmployeeDto {
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/entity/dto/EmployeeDto.java:3]: why?

File: src/main/java/com/itechart/students_lab/waybill_suppliers/service/WarehouseService.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    public WarehouseDto findById(Long id) {
+        Warehouse warehouse = warehouseRepo.findById(id).orElseThrow(
+                () -> new EntityNotFoundException(
+                        String.format(WAREHOUSE_WITH_ID_NOT_FOUND, id)));
+        return warehouseMapper.warehouseToWarehouseDto(warehouse);
+    }
+
+    @Transactional
+    public WarehouseDto create(WarehouseDto warehouseDto) {
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/service/WarehouseService.java:55]: 2 calls to db? Make it one please

File: src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java [hunk 5]
```diff
@@ -47,34 +46,22 @@
             message = "Warehouse id must be positive number")
                                          @PathVariable Long id) {
         WarehouseDto warehouse = warehouseService.findById(id);
-        if (warehouse == null) {
-            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
-        }
-        return new ResponseEntity<>(warehouse, HttpStatus.OK);
+        return warehouse == null
+                ? new ResponseEntity<>(HttpStatus.NO_CONTENT)
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/controller/WarehouseController.java:50]: I would personally go with not found 404, but its your call

File: src/main/java/com/itechart/students_lab/waybill_suppliers/repository/WarehouseRepo.java [hunk 6]
```diff
@@ -2,24 +2,9 @@
...
 import java.util.Collection;
-import java.util.List;
 
 public interface WarehouseRepo extends JpaRepository<Warehouse, Long> {
-    @Query("SELECT w FROM Warehouse w WHERE (w.address.state = :addressState "
-            + "AND w.address.city = :addressCity "
-            + "AND w.address.firstAddressLine = :addressLine1 "
-            + "AND w.address.secondAddressLine = :addressLine2) "
-            + "OR (w.name = :whName AND w.customer.name = :whCustomer)")
-    List<Warehouse> findByAddressOrCustomerAndName(@Param("addressState") String addressState,
```
Comment (arramanenka) [inline src/main/java/com/itechart/students_lab/waybill_suppliers/repository/WarehouseRepo.java]: So, you dont need it afterall?

**Review by arramanenka — APPROVED** [review]

File: src/main/java/com/itechart/students_lab/waybill_suppliers/repository/WarehouseRepo.java (hunk 6 above)
Comment (anto-belo) [reply src/main/java/com/itechart/students_lab/waybill_suppliers/repository/WarehouseRepo.java]: The burden of [almost all] validation fell on the db, thus yes, it's not nedeed anymore

**Pull request merged**